src/programming-assistants.rst

```diff
@@ -1,23 +1,179 @@
+4. New fused completions - The tool is able to make completions on sequences and methods it has not been trained on before.
+
+The evaluation of the tools proved it to be very fast and effective. Out of the 84 examples tested on, the desired completion appeared in the top 3 results in 90% of the cases. More details on the implementation of the model and the evaluation metrics used can be found `here <http://www.cs.technion.ac.il/~yahave/papers/pldi14-statistical.pdf>`_.
+
+Pythia, an AI-assisted code completion system, is another Microsoft research project that has since been deployed to Visual Studio IntelliCode :cite:`svyatkovskiy2019pythia`. Pythia not only suggests method completions but also search for relevant APIs for suggestion. Pythia also uses abstract syntax trees as input to train large-scale deep learning models and attempts to predict completions within 100 ms. To compare the perform of the Pythia model, the researchers also generated models using the frequency approach and Markov-Chain approach.
+
+The frequency baseline model was generated by training a model based on the occurrence count of methods in a corpus. It was further improved by providing additional context information concerning whether or not the code was inside an "if" statement. The Markov-Chain model was model trained to determine the next method given a sequence of prior commands for certain Python libraries. For example:
+
+.. code-block:: python
+  os.path.isfile -> os.remove -> ?
+
+After creating the baseline models, a long short-term memory model was created with backpropagation through time, a gradient-based neural network training algorithm, applied to the model. More implementation details can be found `here <http://delivery.acm.org/10.1145/3340000/3330699/p2727-svyatkovskiy.pdf?ip=35.1.173.243&id=3330699&acc=OPENTOC&key=93447E3B54F7D979%2E0A17827594E6F2C8%2E4D4702B0C3E38B35%2E9F04A3A78F7D3B8D&__acm__=1571279087_e03a44d506330ae0e8a0d76b148d8a73>`_. Comparing all models created in the study, the frequency and Markov-Chain baselines were 67% and 83% accurate in predicting the expected completion within the top 5 results. The Pythia model was able to out perfom both models with an accuracy of 92% for the same metric, giving the researchers enough evidence to publish and integrate the tool into Visual Studio.
+
 Program Repair
 ==============
 
+**Program Repair** refers to the process of automatic repairing of a code corpus that contains syntactical and/or logical errors. It is intended to speed up the time and effort exhausted to debug a software project. Debugging can consume a significant amount of time the larger or more complex the project is. Not only does the root cause of an issue have to be found but the bug itself has to be fixed making the whole process very tedious. Some common techniques used in Program Repair are statistical fault localization and component-based program synthesis.
```

Comment:
Program repair is about fixing type errors or run-time errors, manifested as exceptions or failed tests. Syntax errors and logical errors that don't lead to type errors or exceptions are not typically handled by program repair systems.